airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py

```diff
@@ -1095,6 +1099,30 @@ def add_to_outputs(
             else:
                 # incremental is handled in the SCD SQL already
                 sql = self.add_incremental_clause(sql)
+                if do_deletions:
+                    deletion_hook = Template(
+                        """
+with partitioned_data as (
+    select row_number() over (
+        partition by _airbyte_unique_key
+        order by _airbyte_active_row desc, _airbyte_ab_id
+    ) as _airbyte_row_num,
+    {{ scd_table }}.*
+    from {{ scd_table }}
+    where 1=1
+    {{ incremental_clause }}
+)
+delete from {{ '{{ this }}' }} where _airbyte_unique_key in (
+    select _airbyte_unique_key from partitioned_data
+    where _airbyte_active_row = 0 and _airbyte_row_num = 1
```

Comment:
Why is it necessary to restrict to `_airbyte_row_num = 1`?

Reply:
oh yeah, I think adding `and _airbyte_end_at not null {{ incremental_clause }}` to the subquery would make everything work.

the record might have been updated but not deleted, in which case the SCD table would contain a row with active_row=0. But we don't want to delete it from the final table, so the `row_num = 1` is to filter for the most recent SCD entry (which would have active_row=1). E.g. when I create a record and then update it twice, this is the resulting SCD:
```
edgao_test_replicated=> select id, bar, _airbyte_active_row, _ab_cdc_deleted_at from foo_table_scd where id = 12 order by _ab_cdc_updated_at;
-[ RECORD 1 ]-------+---------
id                  | 12
bar                 | test9
_airbyte_active_row | 0
_ab_cdc_deleted_at  |
-[ RECORD 2 ]-------+---------
id                  | 12
bar                 | test9-1
_airbyte_active_row | 0
_ab_cdc_deleted_at  |
-[ RECORD 3 ]-------+---------
id                  | 12
bar                 | test9-2
_airbyte_active_row | 1
_ab_cdc_deleted_at  |
```

(I should probably be adding a lot of comments to this file...)

Reply:
hmmm. I actually think the partitioning might be necessary? For handling the situation where a stream emits multiple records with `null` for their cursor field - in this case, the SCD table could contain some records with active_row=0 and end_at=null, even though a later record has active_row=1. ([example](https://github.com/airbytehq/airbyte/blob/cdb972811dec629d232df1dd30e41624eb64ece7/airbyte-integrations/bases/base-normalization/integration_tests/resources/test_simple_streams/data_input/messages.txt#L24-L25))

Does having the incremental clause inside the `partitioned_data` query not prevent scanning the whole SCD table?

Reply:
On BigQuery for example, the partitions of `*_scd` tables are using `_airbyte_active_row` column for partitions (and some other columns for clustering):
https://github.com/airbytehq/airbyte/blob/cdb972811dec629d232df1dd30e41624eb64ece7/airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/bigquery/test_simple_streams/models/generated/airbyte_incremental/scd/test_normalization/dedup_exchange_rate_scd.sql#L2-L3

So if the where clauses (in the incremental_clause) are referring to these columns, BQ will be able to prune partitions/indexes and avoid full table scans yes.

In conclusion, what I initially thought about:
```
delete from {{ '{{ this }}' }} where _airbyte_unique_key in ( select _airbyte_unique_key from {{ scd_table }} where _airbyte_active_row = 0 and _ab_cdc_deleted_at is not null
```

could be inefficient (the partition `_airbyte_active_row = 0` is much larger than `_airbyte_active_row = 1` over time)

Reply:
BTW would it work to do a pre-hook delete on the final table instead of a post-hook?

Isn't that easier to handle?

Reply:
Did you try to run the deletes from the final table as a post hook of the scd table too?

Reply:
if you struggle with CTEs, you can make a separate materialized table (or view) instead of CTE (similar to the *_ab4 models that you can then use in the delete statement, etc)